wagtail/users/views/users.py

```diff
@@ -70,74 +70,71 @@ def get_users_filter_query(q, model_fields):
     return conditions
 
 
-@any_permission_required(add_user_perm, change_user_perm, delete_user_perm)
-@vary_on_headers("X-Requested-With")
-def index(request, *args):
-    q = None
-    is_searching = False
-
-    group = None
-    group_filter = Q()
-    if args:
-        group = get_object_or_404(Group, id=args[0])
-        group_filter = Q(groups=group) if args else Q()
-
-    model_fields = [f.name for f in User._meta.get_fields()]
-
-    if "q" in request.GET:
-        form = SearchForm(request.GET, placeholder=_("Search users"))
-        if form.is_valid():
-            q = form.cleaned_data["q"]
-            is_searching = True
-            conditions = get_users_filter_query(q, model_fields)
-
-            users = User.objects.filter(group_filter & conditions)
-    else:
-        form = SearchForm(placeholder=_("Search users"))
-
-    if not is_searching:
-        users = User.objects.filter(group_filter)
+class Index(IndexView):
+    """
+    Lists the users for management within the admin.
+    """
+
+    any_permission_required = ["add", "change", "delete"]
+    permission_policy = ModelPermissionPolicy(User)
+    model = User
+    context_object_name = "users"
+    index_url_name = "wagtailusers_users:index"
+    add_url_name = "wagtailusers_users:add"
+    edit_url_name = "wagtailusers_users:edit"
+    default_ordering = "name"
+    paginate_by = 20
+    template_name = None
+    is_searchable = True
+    page_title = gettext_lazy("Users")
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        setattr(self, "template_name", self.get_template())
+        self.group = get_object_or_404(Group, id=args[0]) if args else None
+        self.group_filter = Q(groups=self.group) if self.group else Q()
+        self.model_fields = [f.name for f in User._meta.get_fields()]
+
+    def get_queryset(self):
+        if self.is_searching:
+            conditions = get_users_filter_query(self.search_query, self.model_fields)
+            users = User.objects.filter(self.group_filter & conditions)
+        else:
+            users = User.objects.filter(self.group_filter)
 
-    if "last_name" in model_fields and "first_name" in model_fields:
-        users = users.order_by("last_name", "first_name")
+        if self.locale:
+            users = users.filter(locale=self.locale)
 
-    if "ordering" in request.GET:
-        ordering = request.GET["ordering"]
+        if "last_name" in self.model_fields and "first_name" in self.model_fields:
+            users = users.order_by("last_name", "first_name")
 
-        if ordering == "username":
+        if self.get_ordering() == "username":
```

Comment:
A bug has been introduced at this line, the available orderings must also be set for this to work.

See `wagtail/admin/views/generic/models.py` snippet below - the ordering must be a valid ordering based on the result of `get_valid_orderings`.

```python

    def get_valid_orderings(self):
        orderings = []
        for col in self.columns:
            if col.sort_key:
                orderings.append(col.sort_key)
                orderings.append("-%s" % col.sort_key)
        print('valid orderings', orderings)
        return orderings

    def get_ordering(self):
        ordering = self.request.GET.get("ordering", self.default_ordering)
        if ordering not in self.get_valid_orderings():
            ordering = self.default_ordering
        return ordering
```

As we are not using the full `columns` feature we can simply add a method to this class.

```

    def get_valid_orderings(self):
        return ["name", "username"]

```

I will apply this fix - everything else looks good.

Reply:
Got it. Thanks for your review and fixing the bug!